docs/model.rst

```diff
@@ -31,34 +167,329 @@ persistence driver. Use it to define fields of your model::
+         });
+
+      }
+   }
+
+
+Strict Fields
+^^^^^^^^^^^^^
+
 .. php:property:: strict_fields
 
-    By default model will only allow you to operate with values for the fields
-    that have been defined through addField(). If you attempt to get, set or
-    otherwise access the value of any other field that has not been properly
-    defined, you'll get exception. Read more about :php:class:`Field`
+By default model will only allow you to operate with values for the fields
+that have been defined through addField(). If you attempt to get, set or
+otherwise access the value of any other field that has not been properly
+defined, you'll get an exception. Read more about :php:class:`Field`
+
+If you set `strict_fields` to false, then the check will not be performed.
+
+Actions
+-------
+Another common thing to define inside :php:meth:`Model::init()` would be
+a user invokable actions::
+
+   class User extends Model {
+
+      function init() {
+
+         parent::init();
 
-    If you set strict_field to false, then the check will not be performed.
+         $this->addField('name');
+         $this->addField('email');
+         $this->addField('password', ['type'=>'password']);
+
+         $this->addAction('send_new_password');
+
+      }
+
+      function send_new_password()
+      {
+         // .. code here
+
+         $this->save(['password'=> .. ]);
+
+         return 'generated and sent password to '.$m['name'];
+      }
+   }
+
+With a method alone, you can generate and send passwords::
+
+   $user->load(3);
+   $user->send_new_password();
+
+but using `$this->addAction()` exposes that method to the ATK UI wigets,
+so if your admin is using `CRUD`, a new button will be available allowing
+passwords to be generated and sent to the users::
+
+   $app->add('CRUD')->setModel(new User($app->db));
+
+Read more about :php:class:`UserAction`
+
+Hooks
+-----
+Hooks (behaviours) can allow you to define callbacks which would trigger
+when data is loaded, saved, deleted etc. Hooks are typically defined in
+:php:meth:`Model::init()` but will be executed accordingly.
+
+There are countless uses for hooks and even more opportunities to use
+hook by all sorts of extensions.
+
+Validation
+^^^^^^^^^^
+
+Validation is an extensive topic, but the simplest use-case would be through
+a hook::
+
+   $this->addField('name');
+
+   $this->addHook('validate', function($m) {
+      if ($m['name'] == 'C#') {
+         return ['name'=>'No sharp objects are allowed'];
+      }
+   });
+
+Now if you attempt to save object, you will receive :php:class:`ValidationException`::
+
+   $model['name'] = 'Swift';
+   $model->saveAndUnload();      // all good
+
+   $model['name'] = 'C#';
+   $model->saveAndUnload();      // exception here
+
+
+Other Uses
+^^^^^^^^^^
+
+Other uses for model hooks are explained in :ref:`Hooks`
+
+
+Inheritance
+-----------
+ATK Data models are really good for structuring hierarchically. Here is example::
```

Comment:
This is not that type of _hierarchically_ I thought about.
You're talking about extending or sub-typing model classes below. It's not hierarchical structure i guess. Hierarchical would be parent/child relation + relation of class to itself (tree structure).